common/types.go

```diff
@@ -58,6 +61,15 @@ func (h *Hash) UnmarshalJSON(input []byte) error {
 	if length >= 2 && input[0] == '"' && input[length-1] == '"' {
 		input = input[1 : length-1]
 	}
+	// strip "0x" for length check
+	if len(input) > 1 && string(input[:2]) == "0x" {
+		input = input[2:]
+	}
+
+	// validate the length of the input hash
+	if len(input) != HashLength*2 {
+		return hashJsonLengthErr
+	}
 	h.SetBytes(FromHex(string(input)))
```

Comment:
And this is where the issue I raised in my hex refactor PR will kick in. Our code blatantly ignores hex parse errors in `common`, so if I give this method a correct length strings to parse (e.g. 64 `x` characters) it will happily parse it with some very undefined behavior.